modules/audio_coding/codecs/opus/audio_encoder_opus.cc

```diff
@@ -623,18 +618,189 @@ AudioEncoder::EncodedInfo AudioEncoderOpusImpl::EncodeImpl(
+          speech = true;
+        } else if (code == 1) {
+          // Code 1: 2 frames in the packet, each with equal compressed size
+          // If both frames were DTX, we would not reach here.
+          consecutive_dtx_frames_ = 0;
+          speech = true;
+        } else if (code == 2) {
+          // Code 2: 2 frames in the packet, with different compressed sizes
+          int size_of_first_frame = encoded[1];
+          int size_of_second_frame = encoded_bytes - size_of_first_frame - 2;
+          if (size_of_first_frame > 0 && size_of_second_frame > 0) {
+            // The second frame has to be speech.
+            consecutive_dtx_frames_ = 0;
+            speech = true;
+          } else if (size_of_first_frame == 0 && size_of_second_frame > 0) {
+            // Second frame may or may not be DTX refresh.
+            speech = (consecutive_dtx_frames_ + 1) != 20;
+            consecutive_dtx_frames_ = 0;
+          } else if (size_of_first_frame > 0 && size_of_second_frame == 0) {
+            // First frame may or may not be DTX refresh.
+            consecutive_dtx_frames_ = 1;
+          } else {
+            // Both frames are size 0/DTX, should not reach here.
+            consecutive_dtx_frames_ += 2;
+          }
+        } else if (code == 3) {
+          // Code 3: an arbitrary number of frames in the packet
+          bool variable = (encoded[1] & 0x80) == 0x80;
+          bool padding = (encoded[1] & 0x40) == 0x40;
+          int M = encoded[1] & 0x3f;
+
+          int padding_header_bytes = 0;
+          int padding_size = 0;
+          if (padding) {
+            if (encoded_bytes < 3) {
+              // The packet should be at least 3 bytes, reset.
+              consecutive_dtx_frames_ = 0;
+              return true;
+            }
+            if (encoded[2] == 0xff) {
+              if (encoded_bytes < 4) {
+                // The packet should be at least 4 bytes, reset.
+                consecutive_dtx_frames_ = 0;
+                return true;
+              }
+              padding_size = 254 + encoded[3];
+              padding_header_bytes = 2;
+            } else {
+              padding_size = encoded[2];
+              padding_header_bytes = 1;
+            }
+          }
+
+          if (variable) {
+            // Frames in the packet have a variable size, a mix of audio and DTX.
+            int offset = 2 + padding_header_bytes;
+            int frame_header_bytes = 0;
+            int total_size_of_frames = 0;
+
+            // Check the worst-case limits to be sure there is enough encoded
+            // data to evaluate.
+            if (encoded_bytes < (size_t)(offset + M * 2)) {
+              // Note: This assumes that actual encoded data is larger than
+              // the guess of two bytes for each header... Reset.
+              consecutive_dtx_frames_ = 0;
+              return true;
+            }
+
+            for (int frame = 0; frame < M - 1; frame++) {
```

Comment:
This is the only time we actually need to walk the packet header, mainly since frame sizes _could be_ larger than 254 bytes and thus need a two-byte header. This only happens when there is a packet with _at least_ one DTX frame and one audio/refresh frame.

Reply:
This could be good to have as a comment in the code too.